src/GafferScene/TransformQuery.cpp

```diff
@@ -0,0 +1,522 @@
+                                io_hash.append( p_scene_plug->fullTransformHash( rpath ) );
+                                invertPlug()->hash( io_hash );
+                            }
+                            break;
+                        }
+                        default:
+                            break;
+                    }
+                }
+            }
+        }
+        else
+        if( ip_plug != 0 )
+        {
+            Gaffer::GraphComponent const* const p_plug_parent = ip_plug->parent();
+
+            if( p_plug_parent == outTranslatePlug() )
+            {
+                ComputeNode::hash( ip_plug, ip_context, io_hash );
+                outMatrixPlug()->hash( io_hash );
+            }
+            else
+            if( p_plug_parent == outRotatePlug() )
+            {
+                ComputeNode::hash( ip_plug, ip_context, io_hash );
+                outMatrixPlug()->hash( io_hash );
+                rotateOrderPlug()->hash( io_hash );
+            }
+            else
+            if( p_plug_parent == outScalePlug() )
+            {
+                ComputeNode::hash( ip_plug, ip_context, io_hash );
+                outMatrixPlug()->hash( io_hash );
+            }
+        }
+    }
+
+    void
+    TransformQuery::compute
+    (
+        Gaffer::ValuePlug* const ip_plug,
+        Gaffer::Context const* const ip_context
+    )
+    const
+    {
+        if( ip_plug == outMatrixPlug() )
+        {
+            Imath::M44f m;
+
+            std::string const loc = locationPlug()->getValue();
+            if( ! loc.empty() )
+            {
+                ScenePlug const* const p_scene_plug = scenePlug();
+                assert( p_scene_plug != 0 );
+
+                ScenePlug::ScenePath const path = ScenePlug::stringToPath( loc );
+
+                if( p_scene_plug->exists( path ) )
+                {
+                    switch( space() )
+                    {
+                        case SpaceLocal:
+                            m = p_scene_plug->transform( path );
+                            break;
+                        case SpaceWorld:
+                            m = p_scene_plug->fullTransform( path );
+                            break;
+                        case SpaceRelative:
+                        {
+                            std::string const rloc = relativeLocationPlug()->getValue();
+                            
+                            if( ! rloc.empty() && ( rloc != loc ) ) // NOTE : If loc == rloc then m is identity
+                            {
+                                Imath::M44f const w2l = p_scene_plug->fullTransform( path );
+                                ScenePlug::ScenePath const rpath = ScenePlug::stringToPath( rloc );
+                                Imath::M44f const r2w = p_scene_plug->fullTransform( rpath ).inverse();
+                                m = r2w * w2l;
```

Comment:
I _think_ there's a bug here, demonstrated by the script below. I'm querying the position of the sphere relative to the camera, and using it to drive the local translate of the cube before parenting it under the camera. I believe that should mean the cube ends up in the same place as the sphere, and it does. But if I then rotate the group, the TransformQuery output changes, even though the relative position of the sphere and the camera stays the same. And the cube drifts off into space...

<details>
```
import Gaffer
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 60, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Sphere"] = GafferScene.Sphere( "Sphere" )
parent.addChild( __children["Sphere"] )
__children["Sphere"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"] = GafferScene.Parent( "Parent" )
parent.addChild( __children["Parent"] )
__children["Parent"]["children"].addChild( GafferScene.ScenePlug( "child1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Cube"] = GafferScene.Cube( "Cube" )
parent.addChild( __children["Cube"] )
__children["Cube"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["TransformQuery"] = GafferScene.TransformQuery( "TransformQuery" )
parent.addChild( __children["TransformQuery"] )
__children["TransformQuery"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Camera"] = GafferScene.Camera( "Camera" )
parent.addChild( __children["Camera"] )
__children["Camera"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Sphere"]["transform"]["translate"].setValue( imath.V3f( 0, 0, -2.99646854 ) )
__children["Sphere"]["radius"].setValue( 0.6100000143051147 )
__children["Sphere"]["__uiPosition"].setValue( imath.V2f( -6.5, 10.9499989 ) )
__children["Group"]["in"][0].setInput( __children["Sphere"]["out"] )
__children["Group"]["in"][1].setInput( __children["Camera"]["out"] )
__children["Group"]["__uiPosition"].setValue( imath.V2f( 4.39915371, -3.8140626 ) )
__children["Parent"]["in"].setInput( __children["Group"]["out"] )
__children["Parent"]["parent"].setValue( '/group/camera' )
__children["Parent"]["children"][0].setInput( __children["Cube"]["out"] )
__children["Parent"]["__uiPosition"].setValue( imath.V2f( 6.64915371, -19.9869881 ) )
__children["Cube"]["transform"]["translate"].setInput( __children["TransformQuery"]["outTranslate"] )
__children["Cube"]["transform"]["rotate"].setInput( __children["TransformQuery"]["outRotate"] )
__children["Cube"]["__uiPosition"].setValue( imath.V2f( 34.9000015, -4.25000048 ) )
__children["TransformQuery"]["scene"].setInput( __children["Group"]["out"] )
__children["TransformQuery"]["location"].setValue( '/group/sphere' )
__children["TransformQuery"]["space"].setValue( 2 )
__children["TransformQuery"]["relativeLocation"].setValue( '/group/camera' )
__children["TransformQuery"]["__uiPosition"].setValue( imath.V2f( 19.4991531, -4.05000019 ) )
__children["Camera"]["__uiPosition"].setValue( imath.V2f( 7.60000086, 11.141139 ) )


del __children
```
</details>

Reply:
6899c37